Register encoding options when adding the service to the service collection

hugohlln opened the pull request:
**Issue:**: #1574

**Description of changes:**
Can now register content type for transformation by passing options to the HostingOptions object

hugohouillon committed:
Can register encoding options when adding the service to the service collection

brignolij commented:
Good idea, Very usefull for webp images with api gateway.

normj commented on Libraries/src/Amazon.Lambda.AspNetCoreServer.Hosting/HostingOptions.cs:
```diff
@@ -12,4 +13,9 @@ public class HostingOptions
+    /// <summary>
+    /// The encoding response options used by the Lambda
+    /// </summary>
+    public IEncodingOptions EncodingOptions { get; set; }
```
`IEncodingOptions` is defined in an `Internal` namespace so it shouldn't be exposed on the public API.

normj commented on Libraries/src/Amazon.Lambda.AspNetCoreServer/Internal/EncodingOptions.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace Amazon.Lambda.AspNetCoreServer.Internal
```
Seems like the `EncodingOptions` and interface should be in `Amazon.Lambda.AspNetCoreServer.Hosting`. They are basically metadata of the `HostingOptions`.

normj commented on Libraries/src/Amazon.Lambda.AspNetCoreServer/Internal/EncodingOptions.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Amazon.Lambda.AspNetCoreServer.Internal
+{
+    public class EncodingOptions : IEncodingOptions
```
We need docs on the class and interface.

normj commented on Libraries/src/Amazon.Lambda.AspNetCoreServer.Hosting/HostingOptions.cs:
```diff
@@ -12,4 +13,9 @@ public class HostingOptions
     public ILambdaSerializer Serializer { get; set; }
+
+    /// <summary>
+    /// The encoding response options used by the Lambda
```
Can you add a sentence that informs the reader that this is used to adjust what content types have their data sent to API Gateway as binary data?

normj reviewed (changes requested):
Thanks for the PR. I think the approach to solve the problem of configuring binary data is good. A few minor comments plus we also need some tests in the `Amazon.Lambda.AspNetCoreServer.Test` project.

hugohouillon committed:
Add comments and change namespace of encoding options

hugohlln commented:
@normj Thanks for your feedbacks

I just dit a new commit with following changes :
- Add comments
- Change the namespace of IEncodingOptions and EncodingOptions

For unit tests, what would be the best way to test the injection of IEncodingOptions ?

CamileDahdah commented on Libraries/src/Amazon.Lambda.AspNetCoreServer.Hosting/Internal/LambdaRuntimeSupportServer.cs:
```diff
@@ -87,9 +87,28 @@ public class APIGatewayHttpApiV2MinimalApi : APIGatewayHttpApiV2ProxyFunction
+            public APIGatewayHttpApiV2MinimalApi(IServiceProvider serviceProvider, IEncodingOptions? encodingOptions = null)
                 : base(serviceProvider)
             {
+                if(encodingOptions != null)
```
This code block and the one under method APIGatewayRestApiMinimalApi are identical. Can you create a private method and call it from both methods?

hugohouillon committed:
Moved the registration of IEncodingOptions to LambdaRuntimeSupportServer level

hugohlln commented:
@CamileDahdah 

I just moved the registration of IEncodingOptions to LambdaRuntimeSupportServer level and removed redundant code from each AbstractAspNetCoreFunction<TREQUEST, TRESPONSE> implementations (APIGatewayHttpApiV2MinimalApi, APIGatewayRestApiMinimalApi and ApplicationLoadBalancerMinimalApi)

normj commented:
Closing as this feature was address in PR https://github.com/aws/aws-lambda-dotnet/pull/2267 that exposed all of the extension points from Amazon.Lambda.AspNetCoreServer onto the HostingOptions including `RegisterResponseContentEncodingForContentEncoding`. That PR was released as part of version 1.10.0 of Amazon.Lambda.AspNetCoreServer.Hosting.

@hugohlln I appreciate your effort but it made sense finally address getting all of the extension points over.
